includes/class-wc-payments.php

```diff
@@ -1291,7 +1291,15 @@ public static function register_checkout_gateway( $payment_method_registry ) {
 	 * Handles upgrade routines.
 	 */
 	public static function install_actions() {
-		if ( version_compare( WCPAY_VERSION_NUMBER, get_option( 'woocommerce_woocommerce_payments_version' ), '>' ) ) {
+		$dependency_service = self::$dependency_service;
+
+		// Make sure we have correct class.
+		if ( ! $dependency_service instanceof WC_Payments_Dependency_Service ) {
+			$dependency_service = new WC_Payments_Dependency_Service();
+		}
+
+		// Don't trigger any update plugin hook and don't bump current plugin version if there is a dependencies issues.
+		if ( version_compare( WCPAY_VERSION_NUMBER, get_option( 'woocommerce_woocommerce_payments_version' ), '>' ) && true === $dependency_service->has_valid_dependencies() ) {
 			do_action( 'woocommerce_woocommerce_payments_updated' );
```

Comment:
Updating the comment to explain why we don't want to run the update plugin hook would be good (i.e. it clears the cache and messes with our ability to detect active merchants in our dependency checks)

Do we think it's safe to skip all the updated actions? I count 10 different things hooked into this at the moment. Are there going to be any side effects to not updating the plugin version here?

I wonder if there's a way for us to be more specific and just skip flushing the cache, but even then I wonder if that might break upgrades for some merchants? Might there be a better / different way of detecting active merchants we could look into instead of relying on the cache being there perhaps?

Reply:
I wonder if the additional check `true === $dependency_service->has_valid_dependencies()` really helps here. 

`install_actions` is only called for the `init` hook in `WC_Payments::init`. However, as you see below, to run `install_actions`, `true === $dependency_service->has_valid_dependencies()` has already passed.

https://github.com/Automattic/woocommerce-payments/blob/83f6905673a9891821bac870367a46560a7ebdc3/includes/class-wc-payments.php#L303-L310

Reply:
> I wonder if the additional check `true === $dependency_service->has_valid_dependencies()` really helps here.
> 
> `install_actions` is only called for the `init` hook in `WC_Payments::init`. However, as you see below, to run `install_actions`, `true === $dependency_service->has_valid_dependencies()` has already passed.
> 
> https://github.com/Automattic/woocommerce-payments/blob/83f6905673a9891821bac870367a46560a7ebdc3/includes/class-wc-payments.php#L303-L310

Before the plugin update, this will be 'passed' `false === self::$dependency_service->has_valid_dependencies()` since the account cache is already there and `install_actions` will be executed so this will be executed right after the plugin update which we want to prevent, but I made an error here where I am doing the same check above. I have to check minimum dependencies without account cache flag.

Reply:
> Updating the comment to explain why we don't want to run the update plugin hook would be good (i.e. it clears the cache and messes with our ability to detect active merchants in our dependency checks)
> 
> Do we think it's safe to skip all the updated actions? I count 10 different things hooked into this at the moment. Are there going to be any side effects to not updating the plugin version here?
> 
> I wonder if there's a way for us to be more specific and just skip flushing the cache, but even then I wonder if that might break upgrades for some merchants? Might there be a better / different way of detecting active merchants we could look into instead of relying on the cache being there perhaps?

A doable strategy is to meticulously inspect all actions, and perform null checks on important variables and classes, focusing specifically on actions we anticipate might be executed outside of the WooCommerce Payments context. Currently, we can pinpoint these actions, but projecting forward, this task may become less feasible.

The goal of this pull request is to avoid triggering plugin update-related hooks in the event of missing dependencies, which will keep the plugin active. Once these dependencies are reinstated, the plugin update hook will become operational during the subsequent request, thereby executing the necessary procedures.

From a different angle, we could think about turning off the update or auto-update features to avoid problems like this if some dependencies are missing. Basically, if something is missing, don't allow the plugin to update.